lib/server/server.js

```diff
@@ -131,122 +132,21 @@ function execute(port) {
-    if (!metadata) {
-      next();
-      return;
-    }
-    const language = metadata.language;
-
-    // determine what file to use according to its id
-    let file;
-    if (metadata.original_id) {
-      if (env.translation.enabled && metadata.language !== 'en') {
-        file =
-          CWD + '/translated_docs/' + metadata.language + '/' + metadata.source;
-      } else {
-        file = CWD + '/versioned_docs/' + metadata.source;
-      }
-    } else {
-      if (!env.translation.enabled || metadata.language === 'en') {
-        file =
-          CWD + '/../' + readMetadata.getDocsPath() + '/' + metadata.source;
-      } else {
-        file =
-          CWD + '/translated_docs/' + metadata.language + '/' + metadata.source;
-      }
-    }
-
-    if (!fs.existsSync(file)) {
+    const found = Object.keys(Metadata)
+      .map(id => Metadata[id])
+      .filter(metadata => metadata.permalink === url);
+    if (found.length < 1) {
       next();
       return;
     }
 
-    let rawContent = readMetadata.extractMetadata(fs.readFileSync(file, 'utf8'))
-      .rawContent;
-
-    // generate table of contents if appropriate
-    if (rawContent && rawContent.indexOf(TABLE_OF_CONTENTS_TOKEN) !== -1) {
-      rawContent = insertTableOfContents(rawContent);
-    }
-
-    let latestVersion = env.latestVersion;
-
-    // replace any links to markdown files to their website html links
-    Object.keys(mdToHtml).forEach(function(key, index) {
-      let link = mdToHtml[key];
-      link = link.replace('/en/', '/' + language + '/');
-      link = link.replace(
-        '/VERSION/',
-        metadata.version && metadata.version !== latestVersion
-          ? '/' + metadata.version + '/'
-          : '/'
-      );
-      // replace relative links without "./"
-      rawContent = rawContent.replace(
-        new RegExp('\\]\\(' + key, 'g'),
-        '](' + link
-      );
-      // replace relative links with "./"
-      rawContent = rawContent.replace(
-        new RegExp('\\]\\(\\./' + key, 'g'),
-        '](' + link
-      );
-    });
-
-    // replace any relative links to static assets to absolute links
-    rawContent = rawContent.replace(
-      /\]\(assets\//g,
-      '](' + siteConfig.baseUrl + 'docs/assets/'
-    );
-
-    removeModuleAndChildrenFromCache('../core/DocsLayout.js');
-    const DocsLayout = require('../core/DocsLayout.js');
-
-    let Doc;
-    if (metadata.layout && siteConfig.layouts[metadata.layout]) {
-      Doc = siteConfig.layouts[metadata.layout]({
-        React,
-        MarkdownBlock: require('../core/MarkdownBlock.js'),
-      });
-    }
+    // mdToHtml is a map from a markdown file name to its html link, used to
+    // change relative markdown links that work on GitHub into actual site links
+    const mdToHtml = render.markdown.mdToHtml(Metadata);
 
-    const docComp = (
-      <DocsLayout
-        metadata={metadata}
-        language={language}
-        config={siteConfig}
-        Doc={Doc}>
```

Comment:
Notice `Doc` was here?

Reply:
That is actually the part I am not sure. I agree the two paths must have discrepancies, so at the beginning try to keep the scope limited. So far this 'Doc' thing is the one of the only two I am not sure. Another one is the `removeModuleAndChildrenFromCache('../core/DocsLayout.js');` line in server.js

But these can be managed if do have to be different. Give render method a 'purpose' parameter would do it. Just was trying to avoid it if not necessary.